File: typescript-data-structure-algorithm/namaste-js/src/OOP/DI/EcommerceSample.ts
interface IProduct {
    id: number
    name: string
    price: number
}

export class ShoppingCart {
    private items: IProduct[] = []

    addItems(item: IProduct): void {
        this.items.push(item)
    }

    getItems(): IProduct[] {
        return this.items
    }
}

export class OrderService {
    constructor(private shoppingCart: ShoppingCart) {}

    placeOrder(): void {
        const items = this.shoppingCart.getItems()
        console.log('Order placed with items =>', items)
    }
}

/** Depedency Injection Container */
export class DIContainer {
    private instances: Map<string, any> = new Map()

    register(key: string, instance: any): void {
        this.instances.set(key, instance)
    }

    resolve<T>(key: string): T {
        if (!this.instances.has(key)) {
            throw new Error(`Dependency ${key} is not registered.`)
        }
        return this.instances.get(key)
    }
}

File: data-structure-algo/src/OOP/DependcyInjection/EcommerceSample.cs
namespace data_structure_algo.src.OOP.DependcyInjection
{
    public class IEcommerceProduct
    {
        public int Id { get; set; }
        public required string Name { get; set; }
        public int Price { get; set; }
    }

    public class ShoppingCart
    {
        List<IEcommerceProduct> products = new();

        public void AddItems(IEcommerceProduct product)
        {
            products.Add(product);
        }

        public List<IEcommerceProduct> GetItems()
        {
            return products;
        }
    }

    public class OrderService
    {
        private readonly ShoppingCart _shoppingCart;

        public OrderService(ShoppingCart shoppingCart)
        {
            _shoppingCart = shoppingCart;
        }

        public void PlaceOrder()
        {
            List<IEcommerceProduct> items = _shoppingCart.GetItems();
            Console.WriteLine($"Order placed with items => {items[0].Name}");
        }
    }
}